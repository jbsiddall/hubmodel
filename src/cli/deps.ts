import { R, z } from "../deps.ts";

export { R, z };
