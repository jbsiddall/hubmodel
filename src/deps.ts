import {z} from "https://deno.land/x/zod@v3.22.2/mod.ts"
import axios, {AxiosInstance} from "npm:axios@1.4.0"

// @deno-types="npm:@types/ramda@0.29.3"
import * as R from "npm:ramda@0.29.0"

export {z, axios, R}
export type {AxiosInstance}