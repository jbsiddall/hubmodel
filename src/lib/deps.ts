import { axios, AxiosInstance, R, z } from "../deps.ts";

export { axios, R, z };
export type { AxiosInstance };
