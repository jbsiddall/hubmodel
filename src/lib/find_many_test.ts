import { assertEquals } from "https://deno.land/std@0.201.0/assert/mod.ts";
import { assertSpyCall, assertSpyCalls, returnsNext, spy, stub } from "https://deno.land/std@0.201.0/testing/mock.ts";
import { createHubspotClient, createHubspotClientForTesting } from "../lib/index.ts";
import { afterEach, beforeEach, describe, it } from "https://deno.land/std@0.201.0/testing/bdd.ts";
import { axios, R } from "./deps.ts";
import MockAdapter from "npm:axios-mock-adapter";
import { assertSnapshot } from "https://deno.land/std@0.201.0/testing/snapshot.ts";

// const assertSnapshotBetter = (ctx: Deno.TestContext, value: unknown, tag?: string) => {
//   assertSnapshot(ctx, value, {name:})
// }

const mockedAxios = new MockAdapter(axios);

// const mockedAxios = stub(axios, 'post')

const client = createHubspotClient({ accessToken: "testtoken" });

afterEach(() => {
  mockedAxios.reset();
});

describe("FindMany Select - type checks", () => {
  // checks types for a bug where we could add properties to the select object that didn't exist
  async function _typeCheckOnlySelectClause() {
    // @ts-expect-error only known properties allowed to be selected
    await client.contacts.findMany({ select: { fake_property_that_doesnt_exist: true } });

    // @ts-expect-error only known properties allowed to be selected
    await client.contacts.findMany({ select: { address: true, fake_property_that_doesnt_exist: true } });

    await client.contacts.findMany({
      select: {
        hs_all_accessible_team_ids: true,
        address: true,
        // @ts-expect-error only known properties allowed to be selected
        fake_property_that_doesnt_exist: true,
      },
    });
  }
});

describe("FindManySelect", () => {
  beforeEach(() => {
    mockedAxios.onPost().replyOnce(200, {
      results: [
        {
          "id": "123",
          "properties": {
            "email": "fakeemail",
            "unwanted": "unwanted",
          },
          "createdAt": "2023-09-15T23:36:11.779Z",
          "updatedAt": "2023-09-15T23:36:11.779Z",
          "archived": false,
        },
      ],
    });
  });

  const f = () => client.contacts.findMany({ select: { email: true } });

  it(async function history(ctx) {
    await f();
    const request = mockedAxios.history.post[0];

    assertSnapshot(
      ctx,
      R.pick(
        ["baseURL", "url", "data", "headers"],
        request,
      ),
    );
  });

  it(async function response(ctx) {
    const contacts = await f();
    const snapshotContacts = contacts.map((c) => {
      return { ...c, updatedAt: undefined };
    });
    assertSnapshot(ctx, snapshotContacts);
  });
});

describe("FindMany Where", () => {
  async function _typeCheckOnlyWhereClause() {
    await client.contacts.findMany({ where: { email: { equals: "fakeemail" } } });

    /** String Property */
    // @ts-expect-error error
    await client.contacts.findMany({ where: { email: { equals: 123 } } });

    await client.contacts.findMany({ where: { email: { equals: null } } });

    // @ts-expect-error error
    await client.contacts.findMany({ where: { email: { equals: new Date() } } });

    /** number Property */
    await client.contacts.findMany({ where: { followercount: { equals: 123 } } });

    // @ts-expect-error error
    await client.contacts.findMany({ where: { followercount: { equals: "123" } } });

    await client.contacts.findMany({ where: { followercount: { equals: null } } });

    await client.contacts.findMany({ where: { email: { not: "fakeemail" } } });
    // @ts-expect-error error
    await client.contacts.findMany({ where: { email: { not: 123 } } });
    await client.contacts.findMany({ where: { email: { not: null } } });
  }
});
