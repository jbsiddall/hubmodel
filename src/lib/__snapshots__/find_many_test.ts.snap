export const snapshot = {};

snapshot[`global > FindManySelect > history 1`] = `
{
  baseURL: "https://api.hubapi.com/",
  data: '{"properties":["email"]}',
  headers: AxiosHeaders {
    "Content-Type": "application/json",
    Accept: "application/json, text/plain, */*",
    Authorization: "Bearer testtoken",
  },
  url: "/crm/v3/objects/contacts/search",
}
`;

snapshot[`global > FindManySelect > response 1`] = `
[
  {
    archived: false,
    createdAt: 2023-09-15T23:36:11.779Z,
    id: "123",
    properties: {
      email: "fakeemail",
    },
    updatedAt: undefined,
  },
]
`;
